Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,393 @@
+    }
+
+    override func layoutAttributesForElementsInRect(rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        var allAttributes = [UICollectionViewLayoutAttributes]()
+        for i in 0 ..< cellCount {
+            let indexPath = NSIndexPath(forRow: i, inSection: 0)
+            let attr = self.computeLayoutAttributesForCellAtIndexPath(indexPath)
+            allAttributes.append(attr)
+        }
+        return allAttributes
+    }
+
+    override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
+        return self.computeLayoutAttributesForCellAtIndexPath(indexPath)
+    }
+
+    override func shouldInvalidateLayoutForBoundsChange(newBounds: CGRect) -> Bool {
+        if newBounds.width < 320 {
+            return false
+        } else {
+            return true
+        }
+    }
+
+    func computeLayoutAttributesForCellAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes {
+        let row = indexPath.row
+        let bounds = self.collectionView!.bounds
+
+        let verticalItemsCount = maxVerticalItemsCount()
+        let horizontalItemsCount = maxHorizontalItemsCount()
+
+
+        let itemsPerPage = verticalItemsCount * horizontalItemsCount
+
+        let columnPosition = row % horizontalItemsCount
+        let rowPosition = (row / horizontalItemsCount) % verticalItemsCount
+        let itemPage = Int(floor(Double(row)/Double(itemsPerPage)))
+
+        let attr = UICollectionViewLayoutAttributes(forCellWithIndexPath: indexPath)
+
+        var frame = CGRectZero
+        frame.origin.x = CGFloat(itemPage) * bounds.size.width + CGFloat(columnPosition) * (itemSize.width + insets.left) + insets.left
+        frame.origin.y = CGFloat(rowPosition) * (itemSize.height + insets.top) + insets.top
+        frame.size = itemSize
+        attr.frame = frame
+        
+        return attr
+    }
+}
+
+/*
+    Defines the number of items to show in topsites for different size classes.
+*/
+struct ASTopSiteSourceUX {
+    static let verticalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 1, UIUserInterfaceSizeClass.Regular : 2]
+    static let horizontalItemsForTraitSizes = [UIUserInterfaceSizeClass.Compact : 3, UIUserInterfaceSizeClass.Regular : 5]
+    static let maxNumberOfPages = 2
+    static let CellIdentifier = "TopSiteCell"
+}
+
+protocol ASHorizontalScrollDelegate {
+    func numberOfVerticalItems(estimatedItems: Int) -> Int
+    func numberOfHorizontalItems(estimatedItems: Int) -> Int
+}
+
+/*
+ This Delegate/DataSource is used to manage the ASHorizontalScrollCell's UICollectionView. 
+ This is left generic enough for it to be re used for other parts of Activity Stream.
+ */
+
+class ASHorizontalScrollSource: NSObject, UICollectionViewDelegate, UICollectionViewDataSource, ASHorizontalScrollDelegate {
+
+    var content: [TopSiteItem] = []
+    var urlPressedHandler: ((NSURL) -> Void)?
+    var pageChangedHandler: ((CGFloat) -> Void)?
```

Comment:
Not quite sure I understand this. The handlers are here because ASHoriziontalSource is the data source for the topsites. I didnt want to pass the homePanelDelegate to it. 
If you mean why arent the clousures in the ASHorizontalScrollDelegate thats because it is the layout delegate and is only responsible for layout stuff. I'll rename it to ASHorizontalScrollLayoutDelegate